### PR TITLE
WIP - Audio Improvements

### PR DESCRIPTION
**Pull request by TheSpiritXIII**
- Audio output and input are opened and closed when needed.
- There is now a feedback widget that displays how loud you're speaking in the audio settings.
- Microphone volume slider is now fully functioning and max value is 400. You will never be speaking too low ever again!

This PR is WIP because it relies on a small hack - audio output does not open when requested, so I use a timer at 1 ms to open output. Without it, audio fails to open output (incorrectly saying it succeeded) the second time you start a call (ringtones and such still worked).

### CODE REVIEW

File: src/audio/audio.cpp [hunk 1]
```diff
@@ -100,41 +108,58 @@ void Audio::setOutputVolume(float volume)
...
-        return;
-    }
 
-    qDebug() << "suscribing input";
     QMutexLocker lock(audioInLock);
-    if (!userCount++ && alInDev)
+    QMutexLocker locker(audioOutLock);
+    assert(userCount >= 0);
+
+    qDebug() << "suscribing input";
```
Comment (antis81) [inline src/audio/audio.cpp:123]: minor typo :)

File: src/audio/audio.cpp [hunk 2]
```diff
@@ -362,9 +420,35 @@ bool Audio::tryCaptureSamples(uint8_t* buf, int framesize)
...
+    }
+
     return true;
 }
 
+void Audio::pauseOutput()
+{
+    QMutexLocker lock(audioOutLock);
+
+    qDebug() << "Puase";
```
Comment (antis81) [inline src/audio/audio.cpp:447]: Minor typo :)

File: src/widget/form/settings/avform.cpp [hunk 3]
```diff
@@ -84,6 +86,7 @@ void AVForm::showEvent(QShowEvent*)
     getAudioInDevices();
     createVideoSurface();
     getVideoDevices();
+    Audio::suscribeInput();
```
Comment (antis81) [inline src/widget/form/settings/avform.cpp:89]: Typo.

File: src/widget/form/settings/avform.cpp [hunk 4]
```diff
@@ -342,10 +352,20 @@ void AVForm::on_playbackSlider_valueChanged(int value)
 
 void AVForm::on_microphoneSlider_valueChanged(int value)
 {
-    Audio::setOutputVolume(value / 100.0);
+    Audio::setInputVolume(value / 100.0);
```
Comment (antis81) [inline src/widget/form/settings/avform.cpp:355]: :+1:

**Comment by antis81** [comment]
This looks mostly good to me :+1:. Can you please correct the typos (is your B key defect? :smile:).

I'm not happy with the mutex in setters/getters, but I'm fine with that, as a future optimization.

File: src/core/core.cpp [hunk 5]
```diff
@@ -82,10 +82,10 @@ Core::Core(QThread *CoreThread, Profile& profile) :
...
 
     // OpenAL init
-    QString outDevDescr = Settings::getInstance().getOutDev();
-    Audio::openOutput(outDevDescr);
-    QString inDevDescr = Settings::getInstance().getInDev();
-    Audio::openInput(inDevDescr);
+    //QString outDevDescr = Settings::getInstance().getOutDev();
+    //Audio::openOutput(outDevDescr);
+    //QString inDevDescr = Settings::getInstance().getInDev();
+    //Audio::openInput(inDevDescr);
```
Comment (agilob) [inline src/core/core.cpp:88]: can we remove that code?

**Comment by antis81** [comment]
Closing -> continued in PR #2415

Thanks for your awesome work @TheSpiritXIII :+1:

**Pull request closed without merging**